controller/api/destination/endpoint_translator.go

```diff
@@ -306,45 +313,52 @@ func toAddr(address watcher.Address) (*net.TcpAddress, error) {
 	}, nil
 }
 
-func toWeightedAddr(address watcher.Address, enableH2Upgrade bool, identityTrustDomain string, controllerNS string, log *logging.Entry) (*pb.WeightedAddr, error) {
-	controllerNSLabel := address.Pod.Labels[k8s.ControllerNSLabel]
-	sa, ns := k8s.GetServiceAccountAndNS(address.Pod)
-	labels := k8s.GetPodLabels(address.OwnerKind, address.OwnerName, address.Pod)
+func toWeightedAddr(address watcher.Address, opaquePorts map[uint32]struct{}, enableH2Upgrade bool, identityTrustDomain string, controllerNS string, log *logging.Entry) (*pb.WeightedAddr, error) {
+	controllerNSLabel := address.Pod.Labels[pkgk8s.ControllerNSLabel]
+	sa, ns := pkgk8s.GetServiceAccountAndNS(address.Pod)
+	labels := pkgk8s.GetPodLabels(address.OwnerKind, address.OwnerName, address.Pod)
 
 	// If the pod is controlled by any Linkerd control plane, then it can be
 	// hinted that this destination knows H2 (and handles our orig-proto
-	// translation) and supports receiving opaque traffic.
+	// translation)
 	var hint *pb.ProtocolHint
 	if enableH2Upgrade && controllerNSLabel != "" {
-		// Get the inbound port from the proxy container's environment
-		// variable so that it can be set in the protocol hint.
-		var inboundPort uint32
-	loop:
-		for _, containerSpec := range address.Pod.Spec.Containers {
-			if containerSpec.Name != k8s.ProxyContainerName {
-				continue
+		hint = &pb.ProtocolHint{
+			Protocol: &pb.ProtocolHint_H2_{
+				H2: &pb.ProtocolHint_H2{},
+			},
+		}
+		opaquePortMatches := false
+		for opaquePort := range opaquePorts {
+			if address.Port == opaquePort {
+				opaquePortMatches = true
+				break
 			}
-			for _, envVar := range containerSpec.Env {
-				if envVar.Name != envInboundListenAddr {
+		}
+		if opaquePortMatches {
+			// Get the inbound port from the proxy container's environment
+			// variable so that it can be set in the protocol hint.
+		loop:
+			for _, containerSpec := range address.Pod.Spec.Containers {
+				if containerSpec.Name != pkgk8s.ProxyContainerName {
 					continue
 				}
-				addr := strings.Split(envVar.Value, ":")
-				port, err := strconv.ParseUint(addr[1], 10, 32)
-				if err != nil {
-					log.Errorf("failed to parse inbound port for proxy container: %s", err)
+				for _, envVar := range containerSpec.Env {
+					if envVar.Name != envInboundListenAddr {
+						continue
+					}
+					addr := strings.Split(envVar.Value, ":")
+					port, err := strconv.ParseUint(addr[1], 10, 32)
+					if err != nil {
+						log.Errorf("failed to parse inbound port for proxy container: %s", err)
+					}
+					hint.OpaqueTransport = &pb.ProtocolHint_OpaqueTransport{
+						InboundPort: uint32(port),
+					}
+					break loop
```

Comment:
tioli: Is it worth splitting this out into a helper like `getInboundPort(pod *Spec) (uint32, error)`?

Reply:
I kind of lean towards leaving it, but I can move it out if that helps! We only use this logic here, so moving it it's own function doesn't remove duplication. Also, if someone is reading through this to understand how this all works and saw `getInboundPort`, I think they'd still need to jump to that definition to see exactly what was happening. So for both of those reasons, I think it's okay being here.

Reply:
:+1: Any time I see labeled loops it makes me think about if breaking out into a separate function is better.

Reply:
Sounds good! I will move it out.